File: app/src/main/java/net/bible/android/view/activity/search/SearchResults.java
/*
 * Copyright (c) 2020 Martin Denham, Tuomas Airaksinen and the And Bible contributors.
 *
 * This file is part of And Bible (http://github.com/AndBible/and-bible).
 *
 * And Bible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * And Bible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with And Bible.
 * If not, see http://www.gnu.org/licenses/.
 *
 */

package net.bible.android.view.activity.search;

 import android.annotation.SuppressLint;
 import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import net.bible.android.activity.R;
import net.bible.android.control.page.window.ActiveWindowPageManagerProvider;
import net.bible.android.control.search.SearchControl;
import net.bible.android.control.search.SearchResultsDto;
import net.bible.android.view.activity.base.Callback;
import net.bible.android.view.activity.base.Dialogs;
import net.bible.android.view.activity.base.ListActivityBase;
import net.bible.android.view.activity.search.searchresultsactionbar.SearchResultsActionBarManager;

import org.apache.commons.lang3.StringUtils;
import org.crosswire.jsword.book.Book;
import org.crosswire.jsword.passage.Key;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

/** do the search and show the search results
 * 
 * @author Martin Denham [mjdenham at gmail dot com]
 */
public class SearchResults extends ListActivityBase {
	private static final String TAG = "SearchResults";
	
    private SearchResultsDto mSearchResultsHolder;
    
    private List<Key> mCurrentlyDisplayedSearchResults = new ArrayList<>();
    private ArrayAdapter<Key> mKeyArrayAdapter;

    private boolean isScriptureResultsCurrentlyShown = true;
    
	private SearchResultsActionBarManager searchResultsActionBarManager;
	
	private SearchControl searchControl;

	private ActiveWindowPageManagerProvider activeWindowPageManagerProvider;

	private static final int LIST_ITEM_TYPE = android.R.layout.simple_list_item_2;

    public SearchResults() {
		super(R.menu.empty_menu);
		
	}
	
    /** Called when the activity is first created. */
    @SuppressLint("MissingSuperCall")
	@Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState, true);
        Log.i(TAG, "Displaying Search results view");
        setContentView(R.layout.list);

		buildActivityComponent().inject(this);

		searchResultsActionBarManager.registerScriptureToggleClickListener(scriptureToggleClickListener);
		setActionBarManager(searchResultsActionBarManager);

        isScriptureResultsCurrentlyShown = searchControl.isCurrentDefaultScripture();

        if (fetchSearchResults()) {
            // initialise adapters before result population - easier when updating due to later Scripture toggle 
        	mKeyArrayAdapter = new SearchItemAdapter(this, LIST_ITEM_TYPE, mCurrentlyDisplayedSearchResults, searchControl);
            setListAdapter(mKeyArrayAdapter);

			populateViewResultsAdapter();
        }
    }

    /** do the search query and prepare results in lists ready for display
     * 
     */
    private boolean fetchSearchResults() {
    	Log.d(TAG, "Preparing search results");
    	boolean isOk;

    	try {
    		// get search string - passed in using extras so extras cannot be null
            Bundle extras = getIntent().getExtras();
			String searchText = extras.getString(SearchControl.SEARCH_TEXT);
			String searchDocument = extras.getString(SearchControl.SEARCH_DOCUMENT);
			if (StringUtils.isEmpty(searchDocument)) {
				searchDocument = activeWindowPageManagerProvider.getActiveWindowPageManager().getCurrentPage().getCurrentDocument().getInitials();
			}
			mSearchResultsHolder = searchControl.getSearchResults(searchDocument, searchText);
			
			// tell user how many results were returned
			String msg;
			if (mCurrentlyDisplayedSearchResults.size()>=SearchControl.MAX_SEARCH_RESULTS) {
				msg = getString(R.string.search_showing_first, SearchControl.MAX_SEARCH_RESULTS);
			} else {
				msg = getString(R.string.search_result_count, mSearchResultsHolder.size());
			}
			Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
			isOk = true;
    	} catch (Exception e) {
    		Log.e(TAG, "Error processing search query", e);
    		isOk = false;
    		Dialogs.getInstance().showErrorMsg(R.string.error_executing_search, new Callback() {
    			@Override
    			public void okay() {
    				onBackPressed();
    			}
    		});
    	}
    	return isOk;
    }

    /** 
     * Move search results into view Adapter
     */
	private void populateViewResultsAdapter() {
		if (isScriptureResultsCurrentlyShown) {
			mCurrentlyDisplayedSearchResults = mSearchResultsHolder.getMainSearchResults();
		} else {
			mCurrentlyDisplayedSearchResults = mSearchResultsHolder.getOtherSearchResults();
		}
		
		// addAll is only supported in Api 11+
		mKeyArrayAdapter.clear();
		for (Key key : mCurrentlyDisplayedSearchResults) {
			mKeyArrayAdapter.add(key);	
		}
		
	}
    
    @Override
	protected void onListItemClick(ListView l, View v, int position, long id) {
    	try {
    		// no need to call HistoryManager.addHistoryItem() here because PassageChangeMediator will tell HistoryManager a change is about to occur
    		
	    	verseSelected(mCurrentlyDisplayedSearchResults.get(position));
		} catch (Exception e) {
			Log.e(TAG, "Selection error", e);
			Dialogs.getInstance().showErrorMsg(R.string.error_occurred, e);
		}
	}
    
    private void verseSelected(Key key) {
    	Log.i(TAG, "chose:"+key);
    	if (key!=null) {
    		// which doc do we show
			String targetDocInitials = getIntent().getExtras().getString(SearchControl.TARGET_DOCUMENT);
			if (StringUtils.isEmpty(targetDocInitials)) {
				targetDocInitials = activeWindowPageManagerProvider.getActiveWindowPageManager().getCurrentPage().getCurrentDocument().getInitials();
			}
			Book targetBook = getSwordDocumentFacade().getDocumentByInitials(targetDocInitials);

			activeWindowPageManagerProvider.getActiveWindowPageManager().setCurrentDocumentAndKey(targetBook, key);
    		
    		// this also calls finish() on this Activity.  If a user re-selects from HistoryList then a new Activity is created
    		returnToPreviousScreen();
    	}
    }

    /**
     * Handle scripture/Appendix toggle
     */
    private OnClickListener scriptureToggleClickListener = new OnClickListener( ) {
		
		@Override
		public void onClick(View view) {
			isScriptureResultsCurrentlyShown = !isScriptureResultsCurrentlyShown;
			populateViewResultsAdapter();
			mKeyArrayAdapter.
			notifyDataSetChanged();
			searchResultsActionBarManager.setScriptureShown(isScriptureResultsCurrentlyShown);
		}
	};

	@Inject
	void setSearchControl(SearchControl searchControl) {
		this.searchControl = searchControl;
	}

	@Inject
	void setSearchResultsActionBarManager(SearchResultsActionBarManager searchResultsActionBarManager) {
		this.searchResultsActionBarManager = searchResultsActionBarManager;
	}

	@Inject
	void setActiveWindowPageManagerProvider(ActiveWindowPageManagerProvider activeWindowPageManagerProvider) {
		this.activeWindowPageManagerProvider = activeWindowPageManagerProvider;
	}
}

File: app/src/main/java/net/bible/android/view/activity/search/SearchResults.kt
/*
 * Copyright (c) 2020-2026 Martin Denham, Sykerö Software / Tuomas Airaksinen and the AndBible contributors.
 *
 * This file is part of AndBible: Bible Study (http://github.com/AndBible/and-bible).
 *
 * AndBible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * AndBible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with AndBible.
 * If not, see http://www.gnu.org/licenses/.
 */
package net.bible.android.view.activity.search

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.bible.android.activity.R
import net.bible.android.activity.databinding.ListBinding
import net.bible.android.control.link.LinkControl
import net.bible.android.control.page.window.WindowControl
import net.bible.android.control.search.SearchControl
import net.bible.android.view.activity.base.Dialogs
import net.bible.android.view.activity.base.ListActivityBase
import net.bible.android.view.activity.page.MainBibleActivity
import net.bible.android.view.activity.search.searchresultsactionbar.SearchResultsActionBarManager
import net.bible.android.control.search.GroupedSearchResult
import net.bible.android.control.search.MultiSearchResultsDto
import net.bible.service.download.FakeBookFactory
import net.bible.service.common.CommonUtils
import net.bible.service.sword.BookAndKey
import net.bible.service.sword.BookAndKeyList
import net.bible.service.sword.SwordDocumentFacade
import org.crosswire.jsword.book.sword.SwordBook
import org.crosswire.jsword.index.IndexStatus
import org.crosswire.jsword.passage.Key
import javax.inject.Inject

class SearchResults : ListActivityBase(R.menu.empty_menu) {
    private lateinit var binding: ListBinding
    private var mSearchResultsHolder: MultiSearchResultsDto? = null
    private var mCurrentlyDisplayedResults: List<GroupedSearchResult> = ArrayList()
    private var mSearchAdapter: MultiSearchItemAdapter? = null
    private var isScriptureResultsCurrentlyShown = true
    override val integrateWithHistoryManager: Boolean = true

    private var selectedTranslations: List<String> = emptyList()
    private var isStrongsSearch = false
    private var documentSelectorMenuItem: MenuItem? = null

    @Inject lateinit var searchResultsActionBarManager: SearchResultsActionBarManager
    @Inject lateinit var searchControl: SearchControl
    @Inject lateinit var linkControl: LinkControl
    @Inject lateinit var windowControl: WindowControl

    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Log.i(TAG, "Displaying Search results view")
        binding = ListBinding.inflate(layoutInflater)
        setContentView(binding.root)
        buildActivityComponent().inject(this)
        searchResultsActionBarManager.registerScriptureToggleClickListener(scriptureToggleClickListener)
        setActionBarManager(searchResultsActionBarManager)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        isScriptureResultsCurrentlyShown = searchControl.isCurrentlyShowingScripture

        selectedTranslations = intent.getStringArrayListExtra(SearchControl.SELECTED_TRANSLATIONS)
            ?: intent.getStringExtra(SearchControl.SEARCH_DOCUMENT)?.let { listOf(it) }
            ?: emptyList()

        isStrongsSearch = intent.getBooleanExtra(SearchControl.IS_STRONGS_SEARCH, false)

        binding.closeButton.setOnClickListener {
            finish()
        }
        lifecycleScope.launch {
            prepareResults()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.search_results_actionbar_menu, menu)
        documentSelectorMenuItem = menu.findItem(R.id.changeSearchDocuments)
        updateDocumentSelectorTitle()
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            R.id.openResultsInWindow -> {
                openResultsInAWindow()
                true
            }
            R.id.changeSearchDocuments -> {
                lifecycleScope.launch { showDocumentSelector() }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun openResultsInAWindow() {
        val lst = BookAndKeyList()
        for (result in mCurrentlyDisplayedResults) {
            for (match in result.translationMatches) {
                lst.addAll(BookAndKey(match.key, match.book))
            }
        }
        linkControl.showLink(FakeBookFactory.multiDocument, lst)
        finish()
    }

    private fun allBibles(): List<SwordBook> =
        SwordDocumentFacade.bibles.filterIsInstance<SwordBook>().sortedBy { it.abbreviation }

    private fun updateDocumentSelectorTitle() {
        val byInitials = allBibles().associateBy { it.initials }
        val label = selectedTranslations
            .mapNotNull { byInitials[it]?.abbreviation }
            .joinToString(", ")
            .ifEmpty { getString(R.string.choose_translations) }
        documentSelectorMenuItem?.title = label
    }

    private suspend fun showDocumentSelector() {
        val candidates = candidateSearchDocuments(isStrongsSearch, allBibles())
        if (candidates.isEmpty()) return

        val selected = Dialogs.multiselect(
            context = this,
            title = getString(R.string.choose_translations),
            items = candidates,
            itemToString = { book ->
                if (book.indexStatus == IndexStatus.DONE) "${book.abbreviation} - ${book.name}"
                else "${book.abbreviation} - ${book.name} (${getString(R.string.search_index_not_created)})"
            },
            preSelected = { selectedTranslations.contains(it.initials) }
        )
        // Empty result = cancelled / dismissed / nothing checked -> no change.
        if (selected.isEmpty()) return

        selectedTranslations = selected.map { it.initials }
        updateDocumentSelectorTitle()
        persistSelection()

        // An unindexed selected document must be indexed before it can be searched. Carry the full
        // search context to SearchIndex so that after indexing the flow returns to SearchResults
        // (SEARCH_TEXT present) and re-runs with the chosen documents. Without SEARCH_TEXT,
        // SearchIndexProgressStatus falls through to the Search entry screen, which shows
        // "An error has occurred" when the current window's document is itself unindexed.
        val unindexed = selected.filter { it.indexStatus != IndexStatus.DONE }
        if (unindexed.isNotEmpty()) {
            startActivity(Intent(this, SearchIndex::class.java).apply {
                putExtra(SearchControl.SEARCH_DOCUMENT, unindexed.first().initials)
                putExtra(SearchControl.SEARCH_TEXT, intent.getStringExtra(SearchControl.SEARCH_TEXT))
                putExtra(SearchControl.IS_STRONGS_SEARCH, isStrongsSearch)
                putStringArrayListExtra(SearchControl.SELECTED_TRANSLATIONS, ArrayList(selectedTranslations))
            })
            return
        }

        prepareResults()
    }

    /**
     * Remembers the chosen documents so the next search restores them. Strong's find-all uses its
     * own key (a Strong's-enabled subset); normal Bible searches share the manual search screen's
     * key, so the two stay in sync.
     */
    private fun persistSelection() {
        val key = if (isStrongsSearch) SearchControl.STRONGS_SEARCH_TRANSLATIONS_PREF
        else SearchControl.SEARCH_TRANSLATIONS_PREF
        CommonUtils.settings.setString(key, selectedTranslations.joinToString(","))
    }

    private suspend fun prepareResults() {
        withContext(Dispatchers.Main) {
            binding.loadingIndicator.visibility = View.VISIBLE
            binding.empty.visibility = View.GONE
        }
        if (fetchSearchResults()) {
            withContext(Dispatchers.Main) {
                mSearchAdapter = MultiSearchItemAdapter(
                    this@SearchResults,
                    mCurrentlyDisplayedResults,
                    ::onTranslationPillClick
                )
                listAdapter = mSearchAdapter as ListAdapter
                populateViewResultsAdapter()
                listView.setSelection(intent.getIntExtra("listPosition", 0))
            }
        }
        withContext(Dispatchers.Main) {
            binding.loadingIndicator.visibility = View.GONE
            if(listAdapter?.isEmpty == true) {
                binding.empty.visibility = View.VISIBLE
            }
        }
    }

    private fun onTranslationPillClick(book: SwordBook, key: Key) {
        intent.putExtra("listPosition", listView.firstVisiblePosition)
        windowControl.activeWindowPageManager.setCurrentDocumentAndKey(book, key)
        val intent = Intent(this, MainBibleActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
        }
        startActivity(intent)
    }

    private suspend fun fetchSearchResults(): Boolean = withContext(Dispatchers.IO) Main@ {
        Log.i(TAG, "Preparing search results")
        var isOk: Boolean
        try {
            val searchText = intent.getStringExtra(SearchControl.SEARCH_TEXT) ?: ""

            if(linkControl.tryToOpenRef(searchText)) {
                historyTraversal.historyManager.popHistoryItem()
                historyTraversal.historyManager.popHistoryItem()
                finish()
                return@Main false
            }

            Log.i(TAG, "Searching: $searchText in ${selectedTranslations.size} translations")

            mSearchResultsHolder = searchControl.getMultiSearchResults(selectedTranslations, searchText)

            withContext(Dispatchers.Main) {
                val resultCount = mSearchResultsHolder?.size ?: 0
                supportActionBar?.title = getString(R.string.multi_search_results, resultCount, selectedTranslations.size)
                updateDocumentSelectorTitle()
                Toast.makeText(
                    this@SearchResults,
                    getString(R.string.search_result_count, resultCount),
                    Toast.LENGTH_SHORT
                ).show()
            }
            isOk = true
        } catch (e: Exception) {
            Log.e(TAG, "Error processing search query", e)
            isOk = false
            Dialogs.showErrorMsg(R.string.error_executing_search) { onBackPressed() }
        }
        return@Main isOk
    }

    private fun populateViewResultsAdapter() {
        mCurrentlyDisplayedResults = if (isScriptureResultsCurrentlyShown) {
            mSearchResultsHolder?.mainSearchResults ?: emptyList()
        } else {
            mSearchResultsHolder?.otherSearchResults ?: emptyList()
        }
        mSearchAdapter?.clear()
        mSearchAdapter?.addAll(mCurrentlyDisplayedResults)
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        // Clicks are handled by the adapter (verse clicks and expand/collapse)
    }

    private val scriptureToggleClickListener = View.OnClickListener {
        isScriptureResultsCurrentlyShown = !isScriptureResultsCurrentlyShown
        populateViewResultsAdapter()
        mSearchAdapter?.notifyDataSetChanged()
        searchResultsActionBarManager.setScriptureShown(isScriptureResultsCurrentlyShown)
    }

    companion object {
        private const val TAG = "SearchResults"
    }
}
